### PR TITLE
Add ability to add table columns

### LINKED ISSUE #495: Implement adding a new column, resizing columns
## Problem

Users might want to add a new column to their table. Users would also want to resize the column widths.

### Adding a column
The ["Working with Columns" design spec](https://wiki.mathesar.org/en/design/specs/working-with-columns) has a solution for adding a new column, which we need to implement on the frontend.

### Resizing columns
The frontend has a set of display stores. The `columnPosition` store contains the position and width of each column, which is currently defaulted to 160px. For each column, we would need to provide a drag indicator which resizes the column and updates this store.

The `columnPosition` information is currently not persisted anywhere. It will be stored in the local storage in a future PR along with a set of table configurations. For the purpose of this issue, we do not have to handle persisting the state when the table is closed.

## Additional context

- Backend work: #199 
- Design issue: #324 
- Design discussion: #436

### PR DESCRIPTION
**Pull request by pegnott**
Fixes #495  

This code adds functionality to create a new table column.
Resizing is in progress and coming soon

**Technical details**
Added addColumn function to TableData.ts to post to API
Added ui elements for column add in Header.svelte
Propagate events/data changes in TableView.svelte

**Screenshots**
New Add Column drop down UI:
image

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: mathesar_ui/src/sections/table-view/Header.scss
Comment (pavish) [inline mathesar_ui/src/sections/table-view/Header.scss:20]: Please add a newline at the end of all files. This is a convention we follow.

Feel free to raise an issue to add a eslint rule to force this.

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pavish) [inline mathesar_ui/src/sections/table-view/Header.svelte:203]: The condition to check if column with same name exists could be made into a reactive statement, as mentioned in previous comment. 
Reference: https://svelte.dev/docs#3_$_marks_a_statement_as_reactive

```suggestion
            <Button appearance="primary" disabled={!newColumnName?.trim() || isDuplicateColumn} on:click={() => addColumn()}>
```

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pavish) [inline mathesar_ui/src/sections/table-view/Header.svelte:104]: This method seems to be used only for checking if a column with same name exists. This could probably be changed to something like:

```
  function isColumnPresent(name) {
    return columns.data?.some(col => col.name.toLowerCase() === name?.toLowerCase());
  }
  
  // Computed property to identify duplicate column
  $: isDuplicateColumn = isColumnPresent(newColumnName);
```

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pavish) [inline mathesar_ui/src/sections/table-view/Header.svelte:209]: As specified in the previous comments, this statement can also use the same computed property.

```suggestion
          {#if isDuplicateColumn}
```

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pavish) [inline mathesar_ui/src/sections/table-view/Header.svelte:188]: Default value of `closeOnInnerClick` is `false`, and `content` is a class that is already assigned to the dropdown's content by default. 

These values need not be explicitly mentioned when using a component.

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pavish) [inline mathesar_ui/src/sections/table-view/Header.svelte:193]: Instead of text, it would be better to use a plus icon, as per the UX.

File: mathesar_ui/src/sections/table-view/TableView.svelte
Comment (pavish) [inline mathesar_ui/src/sections/table-view/TableView.svelte:127]: Please refrain from using the angle syntax for casting. While this is completely valid, as a convention, we intend to use the new 'as' keyword.

```suggestion
    const newColumn:TableColumn = (event.detail ?? null) as TableColumn;
```

File: mathesar_ui/src/stores/tableData.ts
Comment (pavish) [inline mathesar_ui/src/stores/tableData.ts:541]: I do not understand the need of using `.then` and `.catch` with the `async await` syntax, unless I'm missing something.

This could be updated to:
```
try {
  const column = await postAPI<TableColumn>(`/tables/${id}/columns/`, new_column);
  void fetchTableDetails(db, id);
  void fetchTableRecords(db, id, true);
  ...
} catch(err) {
  // catch block
 ...
}
```

**Review by pavish — CHANGES_REQUESTED** [review]
Hey @pegnott, the add column functionality looks good! :)

I have some review comments, but most are minor and convention based.

File: mathesar_ui/src/sections/table-view/Header.scss
Comment (pegnott) [reply mathesar_ui/src/sections/table-view/Header.scss]: I have checked in the file with newline as specified and have updated my IDE settings.

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pegnott) [reply mathesar_ui/src/sections/table-view/Header.svelte]: I have implemented this as suggested. 
Thanks!

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pegnott) [reply mathesar_ui/src/sections/table-view/Header.svelte]: I have implemented this as suggested.
Thanks!

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pegnott) [reply mathesar_ui/src/sections/table-view/Header.svelte]: I have implemented this as suggested.
Thanks!

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pegnott) [reply mathesar_ui/src/sections/table-view/Header.svelte]: I have removed the unneeded props, thank you!

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pegnott) [reply mathesar_ui/src/sections/table-view/Header.svelte]: Done, thanks!

File: mathesar_ui/src/stores/tableData.ts
Comment (pegnott) [reply mathesar_ui/src/stores/tableData.ts]: Have adjusted as suggested. Thanks!

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pegnott) [reply mathesar_ui/src/sections/table-view/Header.svelte]: https://github.com/pegnott/mathesar/blob/754414d202a29033bd437f2d845269d5885a5831/mathesar_ui/src/sections/table-view/Header.svelte#L241

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pegnott) [reply mathesar_ui/src/sections/table-view/Header.svelte]: https://github.com/pegnott/mathesar/blob/754414d202a29033bd437f2d845269d5885a5831/mathesar_ui/src/sections/table-view/Header.svelte#L149

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pegnott) [reply mathesar_ui/src/sections/table-view/Header.svelte]: https://github.com/pegnott/mathesar/blob/754414d202a29033bd437f2d845269d5885a5831/mathesar_ui/src/sections/table-view/Header.svelte#L245

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pegnott) [reply mathesar_ui/src/sections/table-view/Header.svelte]: https://github.com/pegnott/mathesar/blob/754414d202a29033bd437f2d845269d5885a5831/mathesar_ui/src/sections/table-view/Header.svelte#L223

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pegnott) [reply mathesar_ui/src/sections/table-view/Header.svelte]: https://github.com/pegnott/mathesar/blob/754414d202a29033bd437f2d845269d5885a5831/mathesar_ui/src/sections/table-view/Header.svelte#L229

File: mathesar_ui/src/sections/table-view/TableView.svelte
Comment (pegnott) [reply mathesar_ui/src/sections/table-view/TableView.svelte]: https://github.com/pegnott/mathesar/blob/754414d202a29033bd437f2d845269d5885a5831/mathesar_ui/src/sections/table-view/TableView.svelte#L127

File: mathesar_ui/src/stores/tableData.ts
Comment (pegnott) [reply mathesar_ui/src/stores/tableData.ts]: https://github.com/pegnott/mathesar/blob/754414d202a29033bd437f2d845269d5885a5831/mathesar_ui/src/stores/tableData.ts#L536

File: mathesar_ui/src/components/moveable/Resizeable.svelte
Comment (pavish) [inline mathesar_ui/src/components/moveable/Resizeable.svelte:53]: @pegnott The svelte-movable package exports a component. For utilities that involve making changes on existing dom elements (such as popper, portal) , it would be best to use an action.

For more information on actions, please refer https://svelte.dev/docs#use_action. You could also look into popper.ts (within /components/common/actions) file in our code base, which is an action wrapper over [popper](https://popper.js.org/).

For our usecase, it would be best to install only the movable package `npm i moveable`, and create an action over it.

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pavish) [inline mathesar_ui/src/sections/table-view/Header.svelte:164]: This changes the resizer instance from column to column.
Currently, this logic leads to errors and unexpected behaviour while trying to resize columns.

<a href="https://www.loom.com/share/b9955bbc1c3c4fca8bb11aad742661cd"> <p>Issues with resizing columns - Watch Video</p>  </a>

Screenshot 2021-08-17 at 12 58 04 AM


Once we have an action for the movable package, it might be better to:
  * create a separate column cell component, and handle the resizing logic within it.
  * save the column size while resizing instead of waiting until resize completes
  * create the resize instance on column component mount instead of during mouseover, and destroy the resize instance during component destroy

File: mathesar_ui/src/stores/tableData.ts
Comment (pavish) [inline mathesar_ui/src/stores/tableData.ts:531]: This method does not perform anything additionally. We could remove this and rename `calculateColumnPosition` to `updateColumnPosition` instead.

**Review by pavish — CHANGES_REQUESTED** [review]

**Comment by pavish** [comment]
@pegnott The add column functionality looks good to me! Thanks!

Resizing columns do not seem stable. I have added a few review comments on it.

File: mathesar_ui/src/components/moveable/Resizeable.svelte
Comment (pegnott) [reply mathesar_ui/src/components/moveable/Resizeable.svelte:53]: > @pegnott The add column functionality looks good to me! Thanks!
> 
> Resizing columns do not seem stable. I have added a few review comments on it.

Thank you. I am addressing these this afternoon.

File: mathesar_ui/src/sections/table-view/Header.svelte
Comment (pegnott) [reply mathesar_ui/src/sections/table-view/Header.svelte]: Good afternoon. I have addressed the requested changes as follows:

- [x] create a separate column cell component, and handle the resizing logic within it.
(https://github.com/pegnott/mathesar/blob/307418dcc4344b3c9b7d937cc21a0b84e2f360c2/mathesar_ui/src/sections/table-view/HeaderCell.svelte)

- [x] save the column size while resizing instead of waiting until resize completes (https://github.com/pegnott/mathesar/blob/307418dcc4344b3c9b7d937cc21a0b84e2f360c2/mathesar_ui/src/components/common/actions/moveable.ts#L44-L47)

- [x] create the resize instance on column component mount instead of during mouseover (https://github.com/pegnott/mathesar/blob/307418dcc4344b3c9b7d937cc21a0b84e2f360c2/mathesar_ui/src/components/common/actions/moveable.ts#L42) 

- [x] destroy the resize instance during component destroy (https://github.com/pegnott/mathesar/blob/307418dcc4344b3c9b7d937cc21a0b84e2f360c2/mathesar_ui/src/components/common/actions/moveable.ts#L35)

**Comment by kgodey** [comment]
@pavish @pegnott How do you want to handle splitting up this PR?

**Comment by pavish** [comment]
@kgodey @pegnott My suggestion would be to create a new branch from the last commit for 'addding a new column', and raise a PR from that branch.

I assume there weren't additional changes with respect to column addition, after starting work on column resizing feature. If there were, we could do them as new commits on top of the created branch.

**Comment by kgodey** [comment]
@pegnott Would you be willing to create a new branch so that we can get the add column feature merged in as @pavish suggested?

**Comment by pegnott** [comment]
Good evening - am looking at best strategy to split this PR up now, will follow up with any status/questions/etc.
Thanks

**Comment by pegnott** [comment]
> @pegnott Would you be willing to create a new branch so that we can get the add column feature merged in as @pavish suggested?

I have created a new branch  and PR
branches are:

- branch: https://github.com/pegnott/mathesar/tree/add_column (PR: https://github.com/centerofci/mathesar/pull/540)
- branch: https://github.com/pegnott/mathesar/tree/resize_columns (PR: https://github.com/centerofci/mathesar/pull/649)

**Comment by kgodey** [comment]
Thanks @pegnott!

**Review by pavish — APPROVED** [review]

**Comment by pavish** [comment]
@pegnott Thank you so much for splitting the PRs. This looks great!

I have approved the changes. I am currently half way through https://github.com/centerofci/mathesar/issues/651, which refactors a lot of areas this PR touches. I am blocking this PR, until that is merged.

Once the changes for the refactor is merged, I'll add an additional commit to this PR, and merge this.

@kgodey I hope this is okay.

**Comment by kgodey** [comment]
@pavish that's fine, thanks for the update!

**Comment by pavish** [comment]
@kgodey I've added additional commits to handle merge conflicts with the refactor. I'm going to go ahead and merge this, since this was previously reviewed. Please feel free to add comments and we can take it up as separate issues.

@ghislaineguerin This functionality differs quite a bit from the UX. We will have to reevaluate this as well.

**Pull request merged**
